subsys/task_wdt/task_wdt.c

```diff
@@ -143,6 +143,13 @@ int task_wdt_init(const struct device *hw_wdt)
 
 	k_timer_init(&timer, task_wdt_trigger, NULL);
 
+#ifdef CONFIG_TASK_WDT_HW_FALLBACK
+	if (hw_wdt_channel < 0) {
+		LOG_ERR("hw_wdt install timeout failed: %d", hw_wdt_channel);
+		return hw_wdt_channel;
+	}
+#endif
+
```

Comment:
I suggest to move this further up, right behind the call to `wdt_install_timeout` into the existing `#ifdef`, so that the context is more clear.

Reply:
@martinjaeger I put the code at that location was so even if the hw_wdt timeout installation failed the task_wdt timer would still be started.  It the case of a misconfigured system this should still provide a level of protection.  I'm happy to move it if you have considered this or if you would prefer me move it and make a second copy of `k_timer_init(&timer, task_wdt_trigger, NULL);` inside the if statement.

Reply:
In that case the started `k_timer` would afterwards continuously try to feed a non-existing hardware watchdog channel ID (negative errno). Also we should probably have the same behavior if a `hw_wdt` device was passed without enabling `CONFIG_TASK_WDT_HW_FALLBACK` then.

I would say that a function call should either be successful or fail completely and in that case clean up any memory or resources it occupied.

This case is a bit special and I agree one could argue that a partial success is more safe, as a non-working watchdog will most likely not be noticed during normal operation (in contrast to other peripherals/subsystems).

@carlescufi What's your opinion on that? Should we allow having a partially successful initialization?

Reply:
I think that, in this particular case, this is acceptable.